docs/sts-migrate-to-private-bucket.md

```diff
@@ -0,0 +1,428 @@
+wget https://raw.githubusercontent.com/mtulio/cloud-credential-operator/doc-sts-updates/docs/sts-oidc-cloudfront.json.tpl
+
+cat sts-oidc-cloudfront.json.tpl \
+   | envsubst \
+   > ${DIR_CCO}/oidc-cloudfront.json
+
+
+export CLOUDFRONT_HOST=$(aws cloudfront create-distribution-with-tags \
+    --distribution-config-with-tags \
+    file://${DIR_CCO}/oidc-cloudfront.json \
+    | jq -r .Distribution.DomainName)
+
+echo ${CLOUDFRONT_HOST}
+echo ${OIDC_BUCKET_HOST}
+```
+
+### Create and patch the new OIDC discovery documents and JWKS<a name="setup-oidc-documents"></a>
+
+- Download the current OIDC files (discovery document and JWKS) to the local disk:
+
+```bash
+aws s3 sync s3://${OIDC_BUCKET_NAME} ./bucket
+```
+
+- Create the new path `/pvt` under the local directory:
+
+```bash
+# update the reference
+mkdir bucket/pvt/
+$ cp -rvf bucket/keys.json bucket/.well-known/ bucket/pvt/
+'bucket/keys.json' -> 'bucket/pvt/keys.json'
+'bucket/.well-known/' -> 'bucket/pvt/.well-known'
+'bucket/.well-known/openid-configuration' -> 'bucket/pvt/.well-known/openid-configuration'
+
+$ ls -a bucket/pvt/
+.  ..  keys.json  .well-known
+
+```
+
+- Patch the new documents with the CloudFront Distribution Domain name:
+```bash
+sed -i "s/${OIDC_BUCKET_HOST}/${CLOUDFRONT_HOST}/g" bucket/pvt/.well-known/openid-configuration
+```
+
+- Upload the patched files to the Bucket with the new object prefix `/pvt`
+
+```bash
+aws s3 sync ./bucket/pvt s3://${OIDC_BUCKET_NAME}/pvt
+```
+
+The new object path, `/pvt`, must be accessed by CloudFront through OAI. The Bucket Policy will be added to allow that operation from CloudFront Distribution.
+
+- Download the existing template to create the Bucket Policy
+
+```bash
+#TODO: replace to master
+#wget https://raw.githubusercontent.com/openshift/cloud-credential-operator/master/docs/sts-oidc-bucket-policy.json.tpl
+wget https://raw.githubusercontent.com/mtulio/cloud-credential-operator/doc-sts-updates/docs/sts-oidc-bucket-policy.json.tpl
+```
+
+- Create the Bucket Policy and apply it
+
+```bash
+cat sts-oidc-bucket-policy.json.tpl \
+   | envsubst \
+   > ${DIR_CCO}/oidc-bucket-policy.json
+
+aws s3api put-bucket-policy \
+    --bucket ${OIDC_BUCKET_NAME} \
+    --policy file://${DIR_CCO}/oidc-bucket-policy.json
+```
+
+Now the CloudFront Distribution must have access to the Bucket object `/pvt/keys.json`, test it:
+
+```
+$ curl https://${CLOUDFRONT_HOST}/keys.json
+```
+
+### Create the new OIDC using CloudFront Distribution<a name="setup-oidc-idp"></a>
+
+- Extract the service account signer public key, to generate the IdP by `ccoctl`:
+
+```bash
+oc get configmap bound-sa-token-signing-certs \
+    --namespace openshift-kube-apiserver \
+    --output json \
+    | jq --raw-output '.data["service-account-001.pub"]' \
+    > serviceaccount-signer.public
+```
+
+- Generate the IdP files into the local directory `new-oidc`:
+
+```bash
+./ccoctl aws create-identity-provider \
```

Comment:
The random Id issue also exists here.
cluster name is jshu-discon
OIDC_BUCKET_HOST is jshu-discon-19238-oidc.s3.us-east-2.amazonaws.com
I changed cluster name to jshu-discon-19238 then commands worked.

Reply:
https://github.com/openshift/cloud-credential-operator/pull/515#discussion_r1130413552